File: public/js/core/reception-cadence.js
// reception-cadence.js — a mandatory, signed, per-tick commitment to
// what a domain has (or has not) received from other domains, and a
// monotonicity check across a domain's own successive claims about the
// same sender. Built from a direct research exchange: what does the
// "common reception cadence" idea actually add, once the parts already
// free from the DAG's own parent-pointer ordering (you cannot reference
// an event as a parent before it exists — already true, no new
// mechanism needed) are correctly set aside?
//
// Two real properties, neither already present anywhere else:
//
// 1. RECURRING COST. R11 (cadence-vdf.js) already makes each domain's
//    own epoch advancement cost real, physically-irreducible sequential
//    compute time — unconditionally, whether connected or not. This
//    file adds a MANDATORY signed commitment at every one of those
//    ticks, empty or full. A Sybil cluster built in private, however
//    patiently, still has to keep signing a real commitment at every
//    real tick to remain outwardly consistent — turning identity
//    maintenance into a recurring, ongoing cost, not the one-time
//    registration burn identity-cost.js already covers. This does NOT
//    prevent a patient attacker from building a real, honest-looking
//    Sybil cluster over real time — nothing here claims to — it only
//    makes doing so cost real, ongoing signing effort for every fake
//    identity, for as long as the fabrication needs to look real.
//
// 2. RECEPTION MONOTONICITY. For a given (A, C) pair, A's own
//    successive reception commitments about C — ordered by A's own
//    real epoch — must reference a non-decreasing position in C's own
//    real, independently-recomputed history. A domain claiming, later
//    in its own real history, to have received an EARLIER state of
//    another domain than it itself already claimed to have seen is a
//    genuine internal inconsistency, catchable by pure recomputation,
//    with no external clock, no position, no propagation-delay bound —
//    exactly the correction that closed the version of this idea that
//    tried to use an external slot anchor instead.
//
// Explicitly NOT what this file does: prove that two domains are
// distinct real-world entities (that is identity-cost.js's job,
// already separate), or prevent a genuinely collaborating pair from
// fabricating a mutually-consistent history together (nothing
// relational, with no external anchor, can ever rule that out — named
// honestly, not glossed over, the same limit this idea's own research
// conversation converged on directly).

/**
 * @typedef {{ sourceDomain: string, eventId: string }} ReceivedRef
 * @typedef {{
 *   domain: string,
 *   epoch: number,
 *   kind: 'empty' | 'full',
 *   receivedFrom: ReceivedRef[],
 * }} ReceptionCommitment
 * @typedef {{
 *   commitments: Record<string, ReceptionCommitment[]>,
 *   maxSeenEpoch: Record<string, Record<string, number>>,
 *   rejections: Array<{ eventId: string, domain: string, reason: string }>,
 * }} ReceptionCadenceState
 */

export function initialReceptionCadenceState() {
  return { commitments: {}, maxSeenEpoch: {}, rejections: [] };
}

function canonicalReceptionMessage({ domain, epoch, kind, receivedFrom }) {
  // Canonical, order-independent: sort receivedFrom deterministically
  // so the same real set always produces the same signed message
  // regardless of local iteration order.
  const sorted = [...receivedFrom].sort((a, b) => (a.sourceDomain + a.eventId).localeCompare(b.sourceDomain + b.eventId));
  return JSON.stringify({ domain, epoch, kind, receivedFrom: sorted });
}

async function verifyCommitmentSignature(payload) {
  const { ed25519 } = await import('@noble/curves/ed25519');
  const { deriveDomainId } = await import('./identity/domain-id.js');
  const fromHex = (hex) => {
    const bytes = new Uint8Array(hex.length / 2);
    for (let i = 0; i < bytes.length; i++) bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    return bytes;
  };
  const message = new TextEncoder().encode(canonicalReceptionMessage(payload));
  let valid;
  try {
    valid = ed25519.verify(fromHex(payload.signature), message, fromHex(payload.signerPubkey));
  } catch {
    return false;
  }
  if (!valid) return false;
  const derivedId = await deriveDomainId(fromHex(payload.signerPubkey));
  return derivedId === payload.domain;
}

/**
 * Applies one 'reception-commit' event. `sourceEpochLookup(sourceDomain,
 * eventId)` is a caller-supplied, pure function returning the real
 * epoch (in the source domain's own real, independently-materialized
 * cadence state) that produced `eventId`, or `null` if no such event
 * genuinely exists there — the actual recompute-don't-trust check: a
 * reference to a real but structurally invalid target is always
 * rejected, never merely assumed correct because it was signed.
 *
 * @param {ReceptionCadenceState} state
 * @param {{ id: string, payload: any }} event
 * @param {(sourceDomain: string, eventId: string) => number | null} sourceEpochLookup
 * @returns {Promise<ReceptionCadenceState>}
 */
export async function applyReceptionCommitEvent(state, event, sourceEpochLookup) {
  const payload = event.payload;
  if (!payload || payload.type !== 'reception-commit') return state;

  const { domain, epoch, kind, receivedFrom, signature, signerPubkey } = payload;
  if (typeof domain !== 'string' || !domain) return reject(state, event.id, domain ?? '', 'missing domain');
  if (!Number.isInteger(epoch) || epoch < 1) return reject(state, event.id, domain, 'epoch must be a positive integer');
  if (kind !== 'empty' && kind !== 'full') return reject(state, event.id, domain, "kind must be 'empty' or 'full'");
  if (!Array.isArray(receivedFrom)) return reject(state, event.id, domain, 'receivedFrom must be an array');
  if (kind === 'empty' && receivedFrom.length !== 0) return reject(state, event.id, domain, "kind='empty' requires an empty receivedFrom");
  if (kind === 'full' && receivedFrom.length === 0) return reject(state, event.id, domain, "kind='full' requires a non-empty receivedFrom");

  if (!(await verifyCommitmentSignature({ domain, epoch, kind, receivedFrom, signature, signerPubkey }))) {
    return reject(state, event.id, domain, 'invalid or non-matching signature');
  }

  // Real, per-reference structural check: every claimed reception must
  // resolve to a real event that genuinely exists, at a real epoch, in
  // the source domain's own independently-materialized history.
  const resolvedEpochs = {};
  for (const ref of receivedFrom) {
    if (typeof ref.sourceDomain !== 'string' || typeof ref.eventId !== 'string') {
      return reject(state, event.id, domain, 'malformed receivedFrom entry');
    }
    const sourceEpoch = sourceEpochLookup(ref.sourceDomain, ref.eventId);
    if (sourceEpoch === null || sourceEpoch === undefined) {
      return reject(state, event.id, domain, `claimed reception of '${ref.eventId}' from '${ref.sourceDomain}' does not correspond to any real event there`);
    }
    resolvedEpochs[ref.sourceDomain] = Math.max(resolvedEpochs[ref.sourceDomain] ?? 0, sourceEpoch);
  }

  // RECEPTION MONOTONICITY: for each referenced source domain, this
  // domain's own claimed "how far along I've seen them" must never go
  // backwards across its own successive, real, epoch-ordered claims.
  const priorMax = state.maxSeenEpoch[domain] ?? {};
  for (const [sourceDomain, newMax] of Object.entries(resolvedEpochs)) {
    const previous = priorMax[sourceDomain] ?? 0;
    if (newMax < previous) {
      return reject(state, event.id, domain, `reception monotonicity violated: previously claimed to have seen '${sourceDomain}' up to epoch ${previous}, now claims only epoch ${newMax}`);
    }
  }

  const newMaxSeenForDomain = { ...priorMax };
  for (const [sourceDomain, newMax] of Object.entries(resolvedEpochs)) {
    newMaxSeenForDomain[sourceDomain] = Math.max(newMaxSeenForDomain[sourceDomain] ?? 0, newMax);
  }

  return {
    ...state,
    commitments: { ...state.commitments, [domain]: [...(state.commitments[domain] ?? []), { domain, epoch, kind, receivedFrom }] },
    maxSeenEpoch: { ...state.maxSeenEpoch, [domain]: newMaxSeenForDomain },
  };
}

function reject(state, eventId, domain, reason) {
  return { ...state, rejections: [...state.rejections, { eventId, domain, reason }] };
}

/**
 * Real, DAG-native derivation of `sourceEpochLookup` -- the "epoch that
 * produced this event" a caller needs, computed by recomputation, not
 * self-declared. If `eventId` is itself a real 'cadence' event
 * belonging to `domain`, its own epoch is authoritative; otherwise,
 * walks the event's real ancestor chain (the same bounded-safety walk
 * causal-condition-evaluator.js's own causal-order primitive already
 * uses) looking for the highest-epoch 'cadence' event of `domain` that
 * causally precedes it. Returns null if `eventId` does not genuinely
 * belong to `domain`, or if no real cadence event of that domain is
 * found among its ancestors -- a fabricated reference to a nonexistent
 * or misattributed event is never silently accepted.
 *
 * @param {Array<{ id: string, parents: string[], payload: any }>} orderedEvents
 * @returns {(sourceDomain: string, eventId: string) => number | null}
 */
export function deriveSourceEpochLookup(orderedEvents) {
  const byId = new Map(orderedEvents.map((e) => [e.id, e]));
  const SAFETY_BOUND = 10000;

  return (sourceDomain, eventId) => {
    const target = byId.get(eventId);
    if (!target || target.payload?.domain !== sourceDomain) return null; // genuinely does not exist, or misattributed — the only real rejection case

    if (target.payload?.type === 'cadence' && Number.isInteger(target.payload.epoch)) {
      return target.payload.epoch;
    }

    // Any other real, correctly-attributed event of this domain is a
    // legitimate fact to reference — identity registration, an early
    // accrual, anything — whether or not the domain has advanced its
    // cadence yet. A real event found with NO cadence ancestors is not
    // "nonexistent"; it is this domain's own real epoch-0 state, the
    // same emergent widening of verifiable data every other real event
    // already gets, with no special first-contact rule needed. Only a
    // fabricated or misattributed reference — caught above — is ever
    // rejected.
    const visited = new Set();
    const queue = [...target.parents];
    let maxEpoch = 0;
    while (queue.length > 0 && visited.size < SAFETY_BOUND) {
      const id = queue.shift();
      if (visited.has(id)) continue;
      visited.add(id);
      const event = byId.get(id);
      if (event) {
        if (event.payload?.type === 'cadence' && event.payload?.domain === sourceDomain && Number.isInteger(event.payload.epoch)) {
          maxEpoch = Math.max(maxEpoch, event.payload.epoch);
        }
        queue.push(...event.parents);
      }
    }
    return maxEpoch; // real event, real domain, real position — 0 is a legitimate epoch-0 state, never treated as absence
  };
}

/**
 * registry(H_d) for reception commitments — mirror of every other
 * materialize* function in this project. `sourceEpochLookup` is
 * threaded through unchanged to every event.
 *
 * @param {Array<{ id: string, payload: any }>} orderedEvents
 * @param {(sourceDomain: string, eventId: string) => number | null} sourceEpochLookup
 * @returns {Promise<ReceptionCadenceState>}
 */
export async function materializeReceptionCadence(orderedEvents, sourceEpochLookup) {
  let state = initialReceptionCadenceState();
  for (const event of orderedEvents) {
    state = await applyReceptionCommitEvent(state, event, sourceEpochLookup);
  }
  return state;
}

File: rust-core/src/reception_cadence.rs
//! reception_cadence.rs — mirror of reception-cadence.js. See that
//! file's own header for the full rationale: a mandatory, signed,
//! per-tick commitment to what a domain has (or has not) received from
//! other domains, and a monotonicity check across a domain's own
//! successive claims about the same sender.
//!
//! Two real properties, neither present anywhere else in this project:
//! (1) recurring cost — a real signed commitment required at every real
//! cadence tick, turning Sybil-cluster maintenance into an ongoing
//! cost, not a one-time registration burn; (2) reception monotonicity —
//! a domain's own successive claims about how far it has seen another
//! domain's real history must never go backwards, checked by pure
//! recomputation, with no external clock, no position, no propagation-
//! delay bound.
//!
//! Explicitly NOT what this file does: prove two domains are distinct
//! real-world entities (identity-cost's job, already separate), or
//! prevent a genuinely collaborating pair from fabricating a mutually-
//! consistent history together — no relational check with no external
//! anchor can ever rule that out, named honestly here as it is in the
//! JS mirror's own header.

use std::collections::HashMap;

use ed25519_dalek::{Signature, Verifier, VerifyingKey};
use serde::{Deserialize, Serialize};

use crate::event::Event;
use crate::identity::derive_domain_id;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct ReceivedRef {
    #[serde(rename = "sourceDomain")]
    pub source_domain: String,
    #[serde(rename = "eventId")]
    pub event_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ReceptionKind {
    Empty,
    Full,
}

#[derive(Debug, Clone)]
pub struct ReceptionCommitment {
    pub domain: String,
    pub epoch: i64,
    pub kind: ReceptionKind,
    pub received_from: Vec<ReceivedRef>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ReceptionRejection {
    pub event_id: String,
    pub domain: String,
    pub reason: String,
}

#[derive(Debug, Clone, Default)]
pub struct ReceptionCadenceState {
    pub commitments: HashMap<String, Vec<ReceptionCommitment>>,
    pub max_seen_epoch: HashMap<String, HashMap<String, i64>>,
    pub rejections: Vec<ReceptionRejection>,
}

/// A pure function resolving a claimed (source_domain, event_id) pair
/// to the real epoch that produced it in that domain's own,
/// independently-materialized cadence state, or None if no such event
/// genuinely exists — the actual recompute-don't-trust check.
pub type SourceEpochLookup<'a> = dyn Fn(&str, &str) -> Option<i64> + 'a;

#[derive(Deserialize)]
struct ReceptionCommitPayload {
    domain: Option<String>,
    epoch: Option<i64>,
    kind: Option<String>,
    #[serde(rename = "receivedFrom", default)]
    received_from: Vec<ReceivedRefRaw>,
    signature: Option<String>,
    #[serde(rename = "signerPubkey")]
    signer_pubkey: Option<String>,
}

#[derive(Deserialize, Serialize, Clone)]
struct ReceivedRefRaw {
    #[serde(rename = "sourceDomain")]
    source_domain: String,
    #[serde(rename = "eventId")]
    event_id: String,
}

fn from_hex(hex: &str) -> Option<Vec<u8>> {
    if hex.len() % 2 != 0 {
        return None;
    }
    (0..hex.len()).step_by(2).map(|i| u8::from_str_radix(&hex[i..i + 2], 16).ok()).collect()
}

/// Canonical message — mirrors canonicalReceptionMessage() in
/// reception-cadence.js byte for byte: JSON.stringify of
/// {domain, epoch, kind, receivedFrom: <sorted>}, with receivedFrom
/// sorted by (sourceDomain + eventId) lexicographically.
fn canonical_reception_message(domain: &str, epoch: i64, kind: &str, received_from: &[ReceivedRefRaw]) -> String {
    let mut sorted = received_from.to_vec();
    sorted.sort_by(|a, b| format!("{}{}", a.source_domain, a.event_id).cmp(&format!("{}{}", b.source_domain, b.event_id)));
    let refs_json: Vec<String> = sorted
        .iter()
        .map(|r| format!("{{\"sourceDomain\":{},\"eventId\":{}}}", serde_json::to_string(&r.source_domain).unwrap(), serde_json::to_string(&r.event_id).unwrap()))
        .collect();
    format!(
        "{{\"domain\":{},\"epoch\":{},\"kind\":{},\"receivedFrom\":[{}]}}",
        serde_json::to_string(domain).unwrap(),
        epoch,
        serde_json::to_string(kind).unwrap(),
        refs_json.join(",")
    )
}

fn verify_commitment_signature(domain: &str, epoch: i64, kind: &str, received_from: &[ReceivedRefRaw], signature_hex: &str, signer_pubkey_hex: &str) -> bool {
    let (Some(pubkey_bytes), Some(sig_bytes)) = (from_hex(signer_pubkey_hex), from_hex(signature_hex)) else { return false };
    let Ok(pubkey_array): Result<[u8; 32], _> = pubkey_bytes.clone().try_into() else { return false };
    let Ok(sig_array): Result<[u8; 64], _> = sig_bytes.try_into() else { return false };
    let Ok(verifying_key) = VerifyingKey::from_bytes(&pubkey_array) else { return false };
    let signature = Signature::from_bytes(&sig_array);

    let message = canonical_reception_message(domain, epoch, kind, received_from);
    if verifying_key.verify(message.as_bytes(), &signature).is_err() {
        return false;
    }
    derive_domain_id(&pubkey_bytes) == domain
}

impl ReceptionCadenceState {
    pub fn new() -> Self {
        Self::default()
    }

    fn reject(&mut self, event_id: &str, domain: &str, reason: impl Into<String>) {
        self.rejections.push(ReceptionRejection { event_id: event_id.to_string(), domain: domain.to_string(), reason: reason.into() });
    }

    /// Applies one 'reception-commit' event.
    pub fn apply_event(&mut self, event: &Event, source_epoch_lookup: &SourceEpochLookup) {
        let kind_field = event.payload.get("type").and_then(|v| v.as_str());
        if kind_field != Some("reception-commit") {
            return;
        }

        let Ok(p) = serde_json::from_value::<ReceptionCommitPayload>(event.payload.clone()) else { return };
        let domain = p.domain.unwrap_or_default();
        if domain.is_empty() {
            self.reject(&event.id, "", "missing domain");
            return;
        }
        let Some(epoch) = p.epoch.filter(|e| *e >= 1) else {
            self.reject(&event.id, &domain, "epoch must be a positive integer");
            return;
        };
        let Some(kind_str) = p.kind else {
            self.reject(&event.id, &domain, "kind must be 'empty' or 'full'");
            return;
        };
        if kind_str != "empty" && kind_str != "full" {
            self.reject(&event.id, &domain, "kind must be 'empty' or 'full'");
            return;
        }
        if kind_str == "empty" && !p.received_from.is_empty() {
            self.reject(&event.id, &domain, "kind='empty' requires an empty receivedFrom");
            return;
        }
        if kind_str == "full" && p.received_from.is_empty() {
            self.reject(&event.id, &domain, "kind='full' requires a non-empty receivedFrom");
            return;
        }

        let (Some(signature), Some(signer_pubkey)) = (p.signature, p.signer_pubkey) else {
            self.reject(&event.id, &domain, "missing signature or signerPubkey");
            return;
        };
        if !verify_commitment_signature(&domain, epoch, &kind_str, &p.received_from, &signature, &signer_pubkey) {
            self.reject(&event.id, &domain, "invalid or non-matching signature");
            return;
        }

        let mut resolved_epochs: HashMap<String, i64> = HashMap::new();
        for r in &p.received_from {
            match source_epoch_lookup(&r.source_domain, &r.event_id) {
                Some(source_epoch) => {
                    let entry = resolved_epochs.entry(r.source_domain.clone()).or_insert(0);
                    if source_epoch > *entry {
                        *entry = source_epoch;
                    }
                }
                None => {
                    self.reject(&event.id, &domain, format!("claimed reception of '{}' from '{}' does not correspond to any real event there", r.event_id, r.source_domain));
                    return;
                }
            }
        }

        let prior_max = self.max_seen_epoch.get(&domain).cloned().unwrap_or_default();
        for (source_domain, new_max) in &resolved_epochs {
            let previous = prior_max.get(source_domain).copied().unwrap_or(0);
            if *new_max < previous {
                self.reject(&event.id, &domain, format!("reception monotonicity violated: previously claimed to have seen '{source_domain}' up to epoch {previous}, now claims only epoch {new_max}"));
                return;
            }
        }

        let received_from: Vec<ReceivedRef> = p.received_from.iter().map(|r| ReceivedRef { source_domain: r.source_domain.clone(), event_id: r.event_id.clone() }).collect();
        let kind = if kind_str == "empty" { ReceptionKind::Empty } else { ReceptionKind::Full };
        self.commitments.entry(domain.clone()).or_default().push(ReceptionCommitment { domain: domain.clone(), epoch, kind, received_from });

        let entry = self.max_seen_epoch.entry(domain).or_default();
        for (source_domain, new_max) in resolved_epochs {
            let e = entry.entry(source_domain).or_insert(0);
            if new_max > *e {
                *e = new_max;
            }
        }
    }

    /// registry(H_d) for reception commitments — mirror of every other
    /// materialize function in this project.
    pub fn materialize(ordered_events: &[&Event], source_epoch_lookup: &SourceEpochLookup) -> ReceptionCadenceState {
        let mut state = ReceptionCadenceState::new();
        for event in ordered_events {
            state.apply_event(event, source_epoch_lookup);
        }
        state
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ed25519_dalek::{Signer, SigningKey};

    fn to_hex(bytes: &[u8]) -> String {
        bytes.iter().map(|b| format!("{:02x}", b)).collect()
    }

    fn signed_commit_event(id: &str, signing_key: &SigningKey, domain: &str, epoch: i64, kind: &str, received_from: Vec<ReceivedRefRaw>) -> Event {
        let message = canonical_reception_message(domain, epoch, kind, &received_from);
        let signature = signing_key.sign(message.as_bytes());
        let pubkey = signing_key.verifying_key().to_bytes();
        let refs_json: Vec<serde_json::Value> = received_from.iter().map(|r| serde_json::json!({"sourceDomain": r.source_domain, "eventId": r.event_id})).collect();
        Event {
            id: id.to_string(),
            parents: vec![],
            payload: serde_json::json!({
                "type": "reception-commit", "domain": domain, "epoch": epoch, "kind": kind, "receivedFrom": refs_json,
                "signature": to_hex(&signature.to_bytes()), "signerPubkey": to_hex(&pubkey),
            }),
        }
    }

    fn lookup_fn<'a>(map: &'a HashMap<&'a str, HashMap<&'a str, i64>>) -> impl Fn(&str, &str) -> Option<i64> + 'a {
        move |source_domain, event_id| map.get(source_domain)?.get(event_id).copied()
    }

    /// Cross-language parity: this exact message was independently
    /// computed by reception-cadence.js's own canonicalReceptionMessage()
    /// (Node) for the identical inputs, confirmed to match byte for
    /// byte on the first attempt.
    #[test]
    fn cross_language_parity_matches_the_real_js_canonical_message() {
        let refs = vec![
            ReceivedRefRaw { source_domain: "c".to_string(), event_id: "y".to_string() },
            ReceivedRefRaw { source_domain: "c".to_string(), event_id: "x".to_string() },
        ];
        let msg = canonical_reception_message("alice", 5, "full", &refs);
        assert_eq!(msg, r#"{"domain":"alice","epoch":5,"kind":"full","receivedFrom":[{"sourceDomain":"c","eventId":"x"},{"sourceDomain":"c","eventId":"y"}]}"#);
    }

    #[test]
    fn an_empty_commitment_is_accepted() {
        let signing_key = SigningKey::from_bytes(&[7u8; 32]);
        let domain = derive_domain_id(&signing_key.verifying_key().to_bytes());
        let event = signed_commit_event("e1", &signing_key, &domain, 1, "empty", vec![]);
        let mut state = ReceptionCadenceState::new();
        let empty_lookup: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        state.apply_event(&event, &lookup_fn(&empty_lookup));
        assert_eq!(state.rejections.len(), 0);
        assert_eq!(state.commitments[&domain].len(), 1);
    }

    #[test]
    fn kind_empty_with_non_empty_received_from_is_rejected() {
        let signing_key = SigningKey::from_bytes(&[7u8; 32]);
        let domain = derive_domain_id(&signing_key.verifying_key().to_bytes());
        let refs = vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "x".to_string() }];
        let event = signed_commit_event("e1", &signing_key, &domain, 1, "empty", refs);
        // Force the structural mismatch directly, matching the JS test's own hand-built malformed payload.
        let mut event = event;
        event.payload["kind"] = serde_json::json!("empty");
        let mut state = ReceptionCadenceState::new();
        let empty_lookup: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        state.apply_event(&event, &lookup_fn(&empty_lookup));
        assert_eq!(state.rejections.len(), 1);
    }

    #[test]
    fn security_a_forged_signature_is_rejected() {
        let real_key = SigningKey::from_bytes(&[7u8; 32]);
        let attacker_key = SigningKey::from_bytes(&[9u8; 32]);
        let domain = derive_domain_id(&real_key.verifying_key().to_bytes());
        // Attacker signs with their OWN key but claims the victim's domain id.
        let event = signed_commit_event("e1", &attacker_key, &domain, 1, "empty", vec![]);
        let mut state = ReceptionCadenceState::new();
        let empty_lookup: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        state.apply_event(&event, &lookup_fn(&empty_lookup));
        assert_eq!(state.rejections.len(), 1);
        assert!(state.rejections[0].reason.contains("signature"));
    }

    #[test]
    fn a_full_commitment_referencing_a_real_existing_source_event_is_accepted() {
        let signing_key = SigningKey::from_bytes(&[7u8; 32]);
        let domain = derive_domain_id(&signing_key.verifying_key().to_bytes());
        let refs = vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "real1".to_string() }];
        let event = signed_commit_event("e1", &signing_key, &domain, 1, "full", refs);
        let mut c_map = HashMap::new();
        c_map.insert("real1", 5i64);
        let mut lookup_map: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        lookup_map.insert("c", c_map);
        let mut state = ReceptionCadenceState::new();
        state.apply_event(&event, &lookup_fn(&lookup_map));
        assert_eq!(state.rejections.len(), 0);
    }

    #[test]
    fn security_a_forged_nonexistent_reference_is_rejected() {
        let signing_key = SigningKey::from_bytes(&[7u8; 32]);
        let domain = derive_domain_id(&signing_key.verifying_key().to_bytes());
        let refs = vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "never-really-existed".to_string() }];
        let event = signed_commit_event("e1", &signing_key, &domain, 1, "full", refs);
        let mut c_map = HashMap::new();
        c_map.insert("real1", 5i64);
        let mut lookup_map: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        lookup_map.insert("c", c_map);
        let mut state = ReceptionCadenceState::new();
        state.apply_event(&event, &lookup_fn(&lookup_map));
        assert_eq!(state.rejections.len(), 1);
        assert!(state.rejections[0].reason.contains("does not correspond to any real event"));
    }

    #[test]
    fn security_reception_monotonicity_violation_is_rejected() {
        let signing_key = SigningKey::from_bytes(&[7u8; 32]);
        let domain = derive_domain_id(&signing_key.verifying_key().to_bytes());
        let mut c_map = HashMap::new();
        c_map.insert("c-at-50", 50i64);
        c_map.insert("c-at-30", 30i64);
        let mut lookup_map: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        lookup_map.insert("c", c_map);
        let mut state = ReceptionCadenceState::new();

        let e1 = signed_commit_event("e1", &signing_key, &domain, 1, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-at-50".to_string() }]);
        state.apply_event(&e1, &lookup_fn(&lookup_map));
        assert_eq!(state.rejections.len(), 0);

        let e2 = signed_commit_event("e2", &signing_key, &domain, 2, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-at-30".to_string() }]);
        state.apply_event(&e2, &lookup_fn(&lookup_map));
        assert_eq!(state.rejections.len(), 1);
        assert!(state.rejections[0].reason.contains("monotonicity violated"));
    }

    #[test]
    fn reception_monotonicity_accepts_later_or_equal_state() {
        let signing_key = SigningKey::from_bytes(&[7u8; 32]);
        let domain = derive_domain_id(&signing_key.verifying_key().to_bytes());
        let mut c_map = HashMap::new();
        c_map.insert("c-at-30", 30i64);
        c_map.insert("c-at-50", 50i64);
        c_map.insert("c-at-50b", 50i64);
        let mut lookup_map: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        lookup_map.insert("c", c_map);
        let mut state = ReceptionCadenceState::new();

        let e1 = signed_commit_event("e1", &signing_key, &domain, 1, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-at-30".to_string() }]);
        state.apply_event(&e1, &lookup_fn(&lookup_map));
        let e2 = signed_commit_event("e2", &signing_key, &domain, 2, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-at-50".to_string() }]);
        state.apply_event(&e2, &lookup_fn(&lookup_map));
        let e3 = signed_commit_event("e3", &signing_key, &domain, 3, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-at-50b".to_string() }]);
        state.apply_event(&e3, &lookup_fn(&lookup_map));

        assert_eq!(state.rejections.len(), 0);
        assert_eq!(state.max_seen_epoch[&domain]["c"], 50);
    }

    #[test]
    fn monotonicity_tracked_independently_per_source_domain() {
        let signing_key = SigningKey::from_bytes(&[7u8; 32]);
        let domain = derive_domain_id(&signing_key.verifying_key().to_bytes());
        let mut c_map = HashMap::new();
        c_map.insert("c-at-50", 50i64);
        let mut d_map = HashMap::new();
        d_map.insert("d-at-5", 5i64);
        let mut lookup_map: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        lookup_map.insert("c", c_map);
        lookup_map.insert("d", d_map);
        let mut state = ReceptionCadenceState::new();

        let e1 = signed_commit_event("e1", &signing_key, &domain, 1, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-at-50".to_string() }]);
        state.apply_event(&e1, &lookup_fn(&lookup_map));
        let e2 = signed_commit_event("e2", &signing_key, &domain, 2, "full", vec![ReceivedRefRaw { source_domain: "d".to_string(), event_id: "d-at-5".to_string() }]);
        state.apply_event(&e2, &lookup_fn(&lookup_map));

        assert_eq!(state.rejections.len(), 0);
    }

    /// End-to-end: the real two-partition, four-domain scenario,
    /// mirroring the JS suite's own end-to-end test.
    #[test]
    fn end_to_end_honest_reception_accepted_dishonest_rewrite_caught() {
        let signing_key = SigningKey::from_bytes(&[7u8; 32]);
        let domain_a = derive_domain_id(&signing_key.verifying_key().to_bytes());
        let mut c_map = HashMap::new();
        c_map.insert("c-genesis", 1i64);
        c_map.insert("c-epoch-10", 10i64);
        c_map.insert("c-epoch-25", 25i64);
        let mut lookup_map: HashMap<&str, HashMap<&str, i64>> = HashMap::new();
        lookup_map.insert("c", c_map);
        let mut state = ReceptionCadenceState::new();

        let first_contact = signed_commit_event("fc", &signing_key, &domain_a, 1, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-epoch-10".to_string() }]);
        state.apply_event(&first_contact, &lookup_fn(&lookup_map));
        assert_eq!(state.rejections.len(), 0);

        let later_honest = signed_commit_event("later", &signing_key, &domain_a, 2, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-epoch-25".to_string() }]);
        state.apply_event(&later_honest, &lookup_fn(&lookup_map));
        assert_eq!(state.rejections.len(), 0);
        assert_eq!(state.max_seen_epoch[&domain_a]["c"], 25);

        let dishonest_rewrite = signed_commit_event("rewrite", &signing_key, &domain_a, 3, "full", vec![ReceivedRefRaw { source_domain: "c".to_string(), event_id: "c-genesis".to_string() }]);
        state.apply_event(&dishonest_rewrite, &lookup_fn(&lookup_map));
        assert_eq!(state.rejections.len(), 1, "A cannot later claim to have seen only an earlier state of C than it already, honestly, claimed to have seen");
    }
}
